inc/saiicmpecho.h

```diff
@@ -0,0 +1,387 @@
+/**
+ * Copyright (c) 2014 Microsoft Open Technologies, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *    not use this file except in compliance with the License. You may obtain
+ *    a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
+ *    CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
+ *    LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
+ *    FOR A PARTICULAR PURPOSE, MERCHANTABILITY OR NON-INFRINGEMENT.
+ *
+ *    See the Apache Version 2.0 License for specific language governing
+ *    permissions and limitations under the License.
+ *
+ *    Microsoft would like to thank the following companies for their review and
+ *    assistance with these files: Intel Corporation, Mellanox Technologies Ltd,
+ *    Dell Products, L.P., Facebook, Inc., Marvell International Ltd.
+ *
+ * @file    saiicmpecho.h
+ *
+ * @brief   This module defines SAI interface
+ */
+
+#if !defined (__SAIICMPECHO_H_)
+#define __SAIICMPECHO_H_
+
+#include <saitypes.h>
+
+/**
+ * @defgroup SAIICMPECHO SAI - ICMP_ECHO specific public APIs and data structures
+ *
+ * @{
+ */
+
+/**
+ * @brief SAI ICMP_ECHO session state
+ */
+typedef enum _sai_icmp_echo_session_state_t
+{
+    /** ICMP_ECHO Session is in Down */
+    SAI_ICMP_ECHO_SESSION_STATE_DOWN = 0,
+
+    /** ICMP_ECHO Session is Up */
+    SAI_ICMP_ECHO_SESSION_STATE_UP,
+
+} sai_icmp_echo_session_state_t;
+
+/**
+ * @brief Defines the operational status of the ICMP_ECHO session
+ */
+typedef struct _sai_icmp_echo_session_state_notification_t
+{
+    /**
+     * @brief ICMP_ECHO Session id
+     *
+     * @objects SAI_OBJECT_TYPE_ICMP_ECHO_SESSION
+     */
+    sai_object_id_t icmp_echo_session_id;
+
+    /** ICMP_ECHO session state */
+    sai_icmp_echo_session_state_t session_state;
+
+} sai_icmp_echo_session_state_notification_t;
+
+/**
+ * @brief SAI attributes for ICMP_ECHO session
+ */
+typedef enum _sai_icmp_echo_session_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_ICMP_ECHO_SESSION_ATTR_START,
+
+    /**
+     * @brief Hardware lookup valid
+     *
+     * @type bool
+     * @flags CREATE_ONLY
+     * @default true
+     */
+    SAI_ICMP_ECHO_SESSION_ATTR_HW_LOOKUP_VALID = SAI_ICMP_ECHO_SESSION_ATTR_START,
+
+    /**
+     * @brief Virtual Router
+     *
+     * @type sai_object_id_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_VIRTUAL_ROUTER
+     * @condition SAI_ICMP_ECHO_SESSION_ATTR_HW_LOOKUP_VALID == true
+     */
+    SAI_ICMP_ECHO_SESSION_ATTR_VIRTUAL_ROUTER,
+
+    /**
+     * @brief Destination Port
+     *
+     * @type sai_object_id_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_PORT
+     * @condition SAI_ICMP_ECHO_SESSION_ATTR_HW_LOOKUP_VALID == false
+     */
+    SAI_ICMP_ECHO_SESSION_ATTR_PORT,
```

Comment:
This means that session is created between peering links. How do we specify more sessions i.e. monitoring between multiple peering links

Reply:
This means that SONiC has to create multiple echo sessions; one per peering ports.
If thats the requirement then you may want to add the CRM attribute in switch. Something like
SAI_SWITCH_ATTR_AVAILABLE_ICMP_ECHO_SESSIONS read only attribute.
Initial read will give the max and as more sessions as added subsequent reads will give remaining.

Reply:
Added SAI_SWITCH_ATTR_AVAILABLE_ICMP_ECHO_SESSION and SAI_SWITCH_ATTR_MAX_ICMP_ECHO_SESSION.

Reply:
Thank You